src/std/text/json.ss

```diff
@@ -135,46 +141,33 @@
 
 (def (read-json-list port)
   (read-char port)
-  (let lp ((els []))
-    (let (next (read-json-list-next port))
-      (if next
-        (lp (cons next els))
-        (reverse! els)))))
+  (def l (with-list-builder (c)
+           (let lp ()
+             (defvalues (elem next?) (read-json-list-next port))
+             (when next? (c elem) (lp)))))
+  (def f (json-list))
+  (if f (apply f l) l))
```

Comment:
It's better to avoid `apply` here, the list could be bigger than 8192 elements which would fail.
How about just pass the list directly, and use `list->vector` instead of `vector` for the parameter?
This way you can also define the parameter to have a default value of `identity` and avoid the check, just call it.